packages/grafana-ui/src/components/BarChart/BarChart.tsx

```diff
@@ -0,0 +1,310 @@
+import React, { useCallback, useMemo, useRef } from 'react';
+import {
+  compareDataFrameStructures,
+  DataFrame,
+  DefaultTimeZone,
+  formattedValueToString,
+  getFieldDisplayName,
+  getFieldSeriesColor,
+  getFieldColorModeForField,
+  TimeRange,
+  VizOrientation,
+  fieldReducers,
+  reduceField,
+  DisplayValue,
+} from '@grafana/data';
+
+import { VizLayout } from '../VizLayout/VizLayout';
+import { Themeable } from '../../types';
+import { useRevision } from '../uPlot/hooks';
+import { UPlotChart } from '../uPlot/Plot';
+import { UPlotConfigBuilder } from '../uPlot/config/UPlotConfigBuilder';
+import { AxisPlacement, ScaleDistribution } from '../uPlot/config';
+import { useTheme } from '../../themes';
+import { GraphNGLegendEvent, GraphNGLegendEventMode } from '../GraphNG/types';
+import { FIXED_UNIT } from '../GraphNG/GraphNG';
+import { LegendDisplayMode, VizLegendItem } from '../VizLegend/types';
+import { VizLegend } from '../VizLegend/VizLegend';
+
+import { BarChartFieldConfig, BarChartOptions, BarValueVisibility, defaultBarChartFieldConfig } from './types';
+import { BarsOptions, getConfig } from './bars';
+
+/**
+ * @alpha
+ */
+export interface Props extends Themeable, BarChartOptions {
+  height: number;
+  width: number;
+  data: DataFrame;
+  onLegendClick?: (event: GraphNGLegendEvent) => void;
+  onSeriesColorChange?: (label: string, color: string) => void;
+}
+
+/**
+ * @alpha
+ */
+export const BarChart: React.FunctionComponent<Props> = ({
+  width,
+  height,
+  data,
+  orientation,
+  groupWidth,
+  barWidth,
+  showValue,
+  legend,
+  onLegendClick,
+  onSeriesColorChange,
+  ...plotProps
+}) => {
+  if (!data || data.fields.length < 2) {
+    return <div>Missing data</div>;
+  }
+
+  // dominik? TODO? can this all be moved into `useRevision`
```

Comment:
@leeoniya useRevision is a more generic hook that can be uses in different scenarios as well.

Alternative, to make this simpler and reusable in BarChart and TimeSeries panel etc would be to have `useDataFrameStructureCompare` hook, that would return the comparison function for useRevision to use. It would be sth like this:

```ts
// useDataFrameStructureCompare.ts

export function useDataFrameStructureCompare() {
  const compareFrames = useCallback((a?: DataFrame | null, b?: DataFrame | null) => {
    if (a && b) {
      return compareDataFrameStructures(a, b);
    }
    return false;
  }, []);

  return compareFrames;
}
```

then this code in BarChart/TimeSeries panel would change to:

```ts
const compareFrames = useDataFrameStructureCompare();
const configRev = useRevision(data, compareFrames);
```

But me and @mckn think we do not need the revision anymore if we slightly change the way we create the config. Basically move it to joinDataFrames logic when both aligned data and the config would be created removing the need for the config to be memoized. Then we would compare the not by reference but by value and pass either previous(unchanged) or new config to the UPlotChart component. But since the config can contain functions it's gonna require those functions memoization for the compare to give correct results. I hope to start POCing this approach later today/early next week.

Reply:
I added this as a TODO on  https://github.com/grafana/grafana/issues/30741

I don't see any small changes that make what we have any better/different